### PR TITLE
OSD-9131 provide userCA expiry metric

### PR DESCRIPTION
**Pull request by dofinn**
This PR adds a new controller that reconciles against ConfigMaps thats match the predicate criteria.

Using CNO as import, go mod pulls highest available version of CNO. This uses sigs.k8s.io/controller-runtime v0.9.0-beta.0 which breaks operator-sdk v0.18.2

This was resolved by explicity `go get github.com/openshift/cluster-network-operator@release-4.7` which uses sigs.k8s.io/controller-runtime v0.6.3 . This suffices as osdk uses sigs.k8s.io/controller-runtime v0.6.5

Other notable change is operator now watches cluster - `WATCH_NAMESPACE = ""` 

Example metric

```
# HELP cluster_proxy_ca_expiry_timestamp Indicates cluster proxy CA expiry unix timestamp in UTC
# TYPE cluster_proxy_ca_expiry_timestamp gauge
cluster_proxy_ca_expiry_timestamp{name="osd_exporter",subject="O=Default Company Ltd,L=Default City,C=XX"} 1.734086723e+09
```

Additionally, when a PEM is provided that fails validation, the error is partially raised vic ClusterOperatorDegraded by CVO but is not entirely clear as machine-config CO throw a related error that appears to be presented rather then the CNO one. Thus a second metric has been added so the ability to explicitly alert on a metric is provided. 

The below is for an invalid CA

```
# HELP cluster_proxy_ca_valid Indicates if cluster proxy CA valid
# TYPE cluster_proxy_ca_valid gauge
cluster_proxy_ca_valid{name="osd_exporter"} 0
```

This can be utilized by a PrometheusRule to create alerting conditions on different expiry thresholds.

### CODE REVIEW

**Comment by mrbarge** [comment]
From my perspective I gave this a test and it seems to be working well, recognized multiple certs in a bundle, produced the correct expiry timestamp, etc. I'll hold off on a lgtm until any other repository stakeholders have had a chance to weigh in.

File: pkg/controller/configmap/configmap_controller.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	certBundle, _, err := validation.TrustBundleConfigMap(cfgMap)
+	if err != nil {
+		// Reporting of failed validation is already funneled through CVO via ClusterOperatorDegraded. This is to explicity catch this condition
+		// and handle gracefully rather then causing stacktrace.
+		if strings.Contains(err.Error(), "failed to parse certificate PEM") {
+			reqLogger.Info("failed to parse certificate PEM")
+			r.metricsAggregator.SetClusterProxyCAValid(false)
+			reqLogger.Info("metric should be set here meow")
```
Comment (mrbarge) [inline pkg/controller/configmap/configmap_controller.go:115]:
```suggestion
			reqLogger.Info("setting CA valid metric to false")
```

File: pkg/controller/configmap/configmap_controller.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	certBundle, _, err := validation.TrustBundleConfigMap(cfgMap)
+	if err != nil {
+		// Reporting of failed validation is already funneled through CVO via ClusterOperatorDegraded. This is to explicity catch this condition
+		// and handle gracefully rather then causing stacktrace.
+		if strings.Contains(err.Error(), "failed to parse certificate PEM") {
```
Comment (mrbarge) [inline pkg/controller/configmap/configmap_controller.go:112]: In my testing of this PR I was able to induce a different failure which didn't match against this test and never set the metric:

```
{"level":"error","ts":1639987467.7797878,"logger":"controller","msg":"Reconciler error","controller":"configmap-controller","name":"user-ca-bundle","namespace":"openshift-config","error":"failed parsing certificate data from ConfigMap \"user-ca-bundle\": failed to parse certificate: asn1: structure error: length too large","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\tpkg/mod/github.com/go-logr/zapr@v0.4.0/zapr.go:132\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.6.5/pkg/internal/controller/controller.go:246\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.6.5/pkg/internal/controller/controller.go:218\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.6.5/pkg/internal/controller/controller.go:197\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\tpkg/mod/k8s.io/apimachinery@v0.19.7/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\tpkg/mod/k8s.io/apimachinery@v0.19.7/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\tpkg/mod/k8s.io/apimachinery@v0.19.7/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.Until\n\tpkg/mod/k8s.io/apimachinery@v0.19.7/pkg/util/wait/wait.go:90"}
```

Maybe just `failed to parse certificate` is safe enough to match on.

File: pkg/controller/configmap/configmap_controller.go (hunk 1 above)
Comment (dofinn) [reply pkg/controller/configmap/configmap_controller.go:115]: :facepalm: this was from debugging. Will commit your suggestion.

**Pull request merged**